src/Features/CSharp/CodeFixes/Async/CSharpAddAsyncCodeFixProvider.cs

```diff
@@ -62,53 +61,45 @@ protected override string GetDescription(Diagnostic diagnostic, SyntaxNode node,
 
         private static SyntaxNode GetContainingMember(SyntaxNode oldNode)
         {
-            var parenthesizedLambda = oldNode
-                .Ancestors()
-                .FirstOrDefault(n =>
-                    n.IsKind(SyntaxKind.ParenthesizedLambdaExpression));
-
-            if (parenthesizedLambda != null)
-            {
-                return parenthesizedLambda;
-            }
-
-            var simpleLambda = oldNode
-                .Ancestors()
-                .FirstOrDefault(n =>
-                    n.IsKind(SyntaxKind.SimpleLambdaExpression));
-
-            if (simpleLambda != null)
+            foreach (var node in oldNode.Ancestors())
             {
-                return simpleLambda;
+                switch (node.Kind())
+                {
+                    case SyntaxKind.ParenthesizedLambdaExpression:
+                    case SyntaxKind.SimpleLambdaExpression:
+                    case SyntaxKind.AnonymousMethodExpression:
+                    case SyntaxKind.MethodDeclaration:
+                        if (node.TypeSwitch(
+                            (ParenthesizedLambdaExpressionSyntax e) => !HasAsyncKeyword(e),
+                            (SimpleLambdaExpressionSyntax e) => !HasAsyncKeyword(e),
+                            (AnonymousMethodExpressionSyntax e) => !HasAsyncKeyword(e),
+                            (MethodDeclarationSyntax e) => !HasAsyncKeyword(e),
+                            (SyntaxNode e) => false))
+                        {
+                            return node;
+                        }
+                        break;
+                    default:
+                        continue;
```

Comment:
It seems kind of weird to use a switch and then TypeSwitch on them. What's wrong with casting in each case?